File: backend/app/schemas/subscription.py
"""
Subscription and billing Pydantic schemas.
"""

from datetime import datetime
from typing import Any, Dict, Literal, Optional
from uuid import UUID

from pydantic import BaseModel, Field, validator

from app.schemas.base import BaseSchema, TimestampSchema, UUIDSchema


# Plan Schemas
class PlanBase(BaseModel):
    """Base plan schema."""

    plan_name: str = Field(..., max_length=50, description="Plan name")
    price_monthly: int = Field(..., ge=0, description="Monthly price in Japanese Yen")
    features: Dict[str, Any] = Field(default_factory=dict, description="Plan features")
    api_quota_daily: int = Field(default=10, ge=0, description="Daily API quota")
    ai_analysis_quota_daily: int = Field(
        default=5, ge=0, description="Daily AI analysis quota"
    )
    is_active: bool = Field(default=True, description="Whether plan is active")


class PlanCreate(PlanBase):
    """Plan creation schema."""

    pass


class PlanUpdate(BaseModel):
    """Plan update schema."""

    plan_name: Optional[str] = Field(None, max_length=50, description="Plan name")
    price_monthly: Optional[int] = Field(
        None, ge=0, description="Monthly price in Japanese Yen"
    )
    features: Optional[Dict[str, Any]] = Field(None, description="Plan features")
    api_quota_daily: Optional[int] = Field(None, ge=0, description="Daily API quota")
    ai_analysis_quota_daily: Optional[int] = Field(
        None, ge=0, description="Daily AI analysis quota"
    )
    is_active: Optional[bool] = Field(None, description="Whether plan is active")


class Plan(PlanBase, TimestampSchema):
    """Plan response schema."""

    id: int = Field(..., description="Plan ID")

    model_config = {"from_attributes": True}


# Subscription Schemas
class SubscriptionBase(BaseModel):
    """Base subscription schema."""

    plan_id: int = Field(..., description="Plan ID")
    status: Literal["active", "inactive", "cancelled", "expired"] = Field(
        default="active", description="Subscription status"
    )
    current_period_start: datetime = Field(..., description="Current period start")
    current_period_end: datetime = Field(..., description="Current period end")


class SubscriptionCreate(SubscriptionBase):
    """Subscription creation schema."""

    user_id: UUID = Field(..., description="User ID")


class SubscriptionUpdate(BaseModel):
    """Subscription update schema."""

    plan_id: Optional[int] = Field(None, description="Plan ID")
    status: Optional[Literal["active", "inactive", "cancelled", "expired"]] = Field(
        None, description="Subscription status"
    )
    current_period_start: Optional[datetime] = Field(
        None, description="Current period start"
    )
    current_period_end: Optional[datetime] = Field(
        None, description="Current period end"
    )


class Subscription(SubscriptionBase, UUIDSchema, TimestampSchema):
    """Subscription response schema."""

    user_id: UUID = Field(..., description="User ID")
    plan: Optional[Plan] = Field(None, description="Associated plan")

    model_config = {"from_attributes": True}


# Subscription Management Schemas
class SubscriptionUpgrade(BaseModel):
    """Subscription upgrade request schema."""

    new_plan_id: int = Field(..., description="New plan ID")
    payment_method_id: Optional[str] = Field(None, description="Payment method ID")


class SubscriptionDowngrade(BaseModel):
    """Subscription downgrade request schema."""

    new_plan_id: int = Field(..., description="New plan ID")
    effective_date: Optional[datetime] = Field(
        None, description="When downgrade takes effect"
    )


class SubscriptionCancel(BaseModel):
    """Subscription cancellation request schema."""

    reason: Optional[str] = Field(
        None, max_length=500, description="Cancellation reason"
    )
    cancel_at_period_end: bool = Field(
        default=True, description="Cancel at period end or immediately"
    )


# Usage and Quota Schemas
class UsageQuota(BaseModel):
    """User usage quota information."""

    api_quota_daily: int = Field(..., description="Daily API quota")
    api_usage_today: int = Field(..., description="API usage today")
    ai_analysis_quota_daily: int = Field(..., description="Daily AI analysis quota")
    ai_analysis_usage_today: int = Field(..., description="AI analysis usage today")
    quota_reset_at: datetime = Field(..., description="When quota resets")


class SubscriptionWithUsage(Subscription):
    """Subscription with usage information."""

    usage_quota: UsageQuota = Field(..., description="Current usage and quota")


# Billing Schemas
class BillingInfo(BaseModel):
    """Billing information schema."""

    customer_id: Optional[str] = Field(None, description="Payment provider customer ID")
    payment_method_id: Optional[str] = Field(
        None, description="Default payment method ID"
    )
    billing_email: Optional[str] = Field(None, description="Billing email address")
    tax_id: Optional[str] = Field(None, description="Tax ID number")


class Invoice(BaseModel):
    """Invoice schema."""

    id: str = Field(..., description="Invoice ID")
    subscription_id: UUID = Field(..., description="Subscription ID")
    amount: int = Field(..., description="Invoice amount in JPY")
    currency: str = Field(default="JPY", description="Currency")
    status: str = Field(..., description="Invoice status")
    created_at: datetime = Field(..., description="Invoice creation date")
    due_date: datetime = Field(..., description="Invoice due date")
    paid_at: Optional[datetime] = Field(None, description="Payment date")


# Plan Comparison Schema
class PlanComparison(BaseModel):
    """Plan comparison schema for frontend."""

    plans: list[Plan] = Field(..., description="Available plans")
    current_plan_id: Optional[int] = Field(None, description="User's current plan ID")
    recommended_plan_id: Optional[int] = Field(None, description="Recommended plan ID")


# Quota Usage Schemas
class DailyUsage(BaseModel):
    """Daily usage statistics."""

    date: str = Field(..., description="Date in ISO format")
    api_calls: int = Field(..., description="Number of API calls")
    ai_analysis_calls: int = Field(..., description="Number of AI analysis calls")
    avg_response_time_ms: float = Field(
        ..., description="Average response time in milliseconds"
    )
    total_cost_usd: float = Field(..., description="Total cost in USD")


class QuotaUsageSummary(BaseModel):
    """Quota usage summary for multiple days."""

    period_days: int = Field(..., description="Number of days in the summary")
    start_date: str = Field(..., description="Start date in ISO format")
    end_date: str = Field(..., description="End date in ISO format")
    current_quotas: Dict[str, int] = Field(..., description="Current daily quotas")
    usage_by_date: list[DailyUsage] = Field(..., description="Usage statistics by date")
    total_api_calls: int = Field(..., description="Total API calls in period")
    total_ai_analysis_calls: int = Field(
        ..., description="Total AI analysis calls in period"
    )
    total_cost_usd: float = Field(..., description="Total cost in period")


class QuotaCheckResult(BaseModel):
    """Result of quota availability check."""

    has_quota: bool = Field(..., description="Whether user has quota available")
    quota_info: Dict[str, Any] = Field(..., description="Detailed quota information")


# Subscription Analytics Schema
class SubscriptionAnalytics(BaseModel):
    """Subscription analytics for admin."""

    total_subscribers: int = Field(..., description="Total number of subscribers")
    active_subscribers: int = Field(..., description="Active subscribers")
    monthly_revenue: int = Field(..., description="Monthly revenue in JPY")
    churn_rate: float = Field(..., description="Monthly churn rate")
    plan_distribution: Dict[str, int] = Field(..., description="Subscribers per plan")

    model_config = {"from_attributes": True}

File: frontend/src/types/subscription.ts
/**
 * Subscription TypeScript interfaces
 */

import { BaseEntity, Timestamp, UUID } from './base';

// Plan Types
export interface Plan extends BaseEntity {
  id: number;
  plan_name: string;
  price_monthly: number; // in Japanese Yen
  features: Record<string, any>;
  api_quota_daily: number;
  ai_analysis_quota_daily: number;
  is_active: boolean;
}

export interface PlanCreate {
  plan_name: string;
  price_monthly: number;
  features?: Record<string, any>;
  api_quota_daily?: number;
  ai_analysis_quota_daily?: number;
  is_active?: boolean;
}

export interface PlanUpdate {
  plan_name?: string;
  price_monthly?: number;
  features?: Record<string, any>;
  api_quota_daily?: number;
  ai_analysis_quota_daily?: number;
  is_active?: boolean;
}

// Subscription Types
export type SubscriptionStatus = 'active' | 'inactive' | 'cancelled' | 'expired';

export interface Subscription extends BaseEntity {
  user_id: UUID;
  plan_id: number;
  status: SubscriptionStatus;
  current_period_start: Timestamp;
  current_period_end: Timestamp;
  plan?: Plan;
}

export interface SubscriptionCreate {
  plan_id: number;
  payment_method_id?: string;
}

export interface SubscriptionUpdate {
  plan_id?: number;
  status?: SubscriptionStatus;
}

export interface SubscriptionWithPlan extends Subscription {
  plan: Plan;
}

// Usage and Quota Types
export interface UsageQuota {
  api_calls_today: number;
  api_quota_daily: number;
  ai_analysis_today: number;
  ai_analysis_quota_daily: number;
  quota_reset_at: Timestamp;
  api_calls_remaining: number;
  ai_analysis_remaining: number;
  api_quota_percentage: number;
  ai_quota_percentage: number;
}

export interface UsageHistory {
  date: string; // YYYY-MM-DD
  api_calls: number;
  ai_analyses: number;
  cost_usd?: number;
}

// Billing Types
export interface BillingInfo {
  customer_id?: string;
  payment_method_id?: string;
  billing_email?: string;
  next_billing_date?: Timestamp;
  billing_amount?: number; // in JPY
}

export interface Invoice {
  id: string;
  subscription_id: UUID;
  amount: number; // in Japanese Yen
  currency: string;
  status: string;
  created_at: Timestamp;
  due_date?: Timestamp;
  paid_at?: Timestamp;
  invoice_url?: string;
}

// Plan Comparison Types
export interface PlanComparison {
  plans: Plan[];
  current_plan_id?: number;
  recommendations: string[];
}

// Subscription Analytics Types
export interface SubscriptionAnalytics {
  total_subscribers: number;
  active_subscribers: number;
  plan_distribution: Record<string, number>;
  monthly_revenue: number; // in JPY
  churn_rate: number;
  growth_rate: number;
}

// Subscription Form Types
export interface SubscriptionFormData {
  plan_id: number;
  payment_method: 'credit_card' | 'bank_transfer' | 'paypal';
  billing_cycle: 'monthly' | 'yearly';
  auto_renew: boolean;
}

export interface PaymentMethodFormData {
  type: 'credit_card' | 'bank_account';
  card_number?: string;
  expiry_month?: string;
  expiry_year?: string;
  cvv?: string;
  cardholder_name?: string;
  bank_account_number?: string;
  bank_code?: string;
  account_holder_name?: string;
  billing_address: {
    line1: string;
    line2?: string;
    city: string;
    state: string;
    postal_code: string;
    country: string;
  };
}

export interface BillingAddressFormData {
  line1: string;
  line2?: string;
  city: string;
  state: string;
  postal_code: string;
  country: string;
  is_default: boolean;
}

// Subscription State Types (for Redux/Context)
export interface SubscriptionState {
  // Current subscription
  currentSubscription?: SubscriptionWithPlan;
  
  // Available plans
  plans: Plan[];
  
  // Usage and quotas
  usageQuota?: UsageQuota;
  usageHistory: UsageHistory[];
  
  // Billing information
  billingInfo?: BillingInfo;
  invoices: Invoice[];
  
  // Plan comparison
  planComparison?: PlanComparison;
  
  // Payment methods
  paymentMethods: PaymentMethod[];
  defaultPaymentMethod?: PaymentMethod;
  
  // Loading states
  isLoadingSubscription: boolean;
  isLoadingPlans: boolean;
  isLoadingUsage: boolean;
  isLoadingBilling: boolean;
  isLoadingInvoices: boolean;
  isLoadingPaymentMethods: boolean;
  
  // Error states
  subscriptionError?: string;
  plansError?: string;
  usageError?: string;
  billingError?: string;
  invoicesError?: string;
  paymentMethodsError?: string;
  
  // Operation states
  isUpgrading: boolean;
  isDowngrading: boolean;
  isCancelling: boolean;
  isAddingPaymentMethod: boolean;
  isUpdatingBilling: boolean;
  
  // UI state
  selectedPlanId?: number;
  showPlanComparison: boolean;
  showUsageDetails: boolean;
  
  // Last updated timestamps
  lastUpdated: {
    subscription?: Timestamp;
    plans?: Timestamp;
    usage?: Timestamp;
    billing?: Timestamp;
    invoices?: Timestamp;
    paymentMethods?: Timestamp;
  };
}

// Payment Method Types
export interface PaymentMethod {
  id: string;
  type: 'credit_card' | 'bank_account' | 'paypal';
  is_default: boolean;
  
  // Credit card specific
  card_brand?: string;
  card_last4?: string;
  card_exp_month?: number;
  card_exp_year?: number;
  
  // Bank account specific
  bank_name?: string;
  account_last4?: string;
  
  // Common
  billing_address?: BillingAddress;
  created_at: Timestamp;
  updated_at: Timestamp;
}

export interface BillingAddress {
  line1: string;
  line2?: string;
  city: string;
  state: string;
  postal_code: string;
  country: string;
}

// Subscription Events Types
export interface SubscriptionEvent {
  id: UUID;
  subscription_id: UUID;
  event_type: SubscriptionEventType;
  event_data: Record<string, any>;
  created_at: Timestamp;
}

export type SubscriptionEventType = 
  | 'subscription_created'
  | 'subscription_updated'
  | 'subscription_cancelled'
  | 'subscription_renewed'
  | 'plan_upgraded'
  | 'plan_downgraded'
  | 'payment_succeeded'
  | 'payment_failed'
  | 'invoice_created'
  | 'invoice_paid'
  | 'quota_exceeded'
  | 'quota_reset';

// Subscription Notifications Types
export interface SubscriptionNotification {
  id: UUID;
  user_id: UUID;
  notification_type: SubscriptionNotificationType;
  title: string;
  message: string;
  is_read: boolean;
  created_at: Timestamp;
}

export type SubscriptionNotificationType = 
  | 'payment_due'
  | 'payment_failed'
  | 'subscription_expiring'
  | 'quota_warning'
  | 'quota_exceeded'
  | 'plan_upgrade_available'
  | 'billing_update_required';

// Subscription Metrics Types
export interface SubscriptionMetrics {
  user_id: UUID;
  current_period: {
    api_calls: number;
    ai_analyses: number;
    cost_usd: number;
    days_remaining: number;
  };
  historical: {
    total_api_calls: number;
    total_ai_analyses: number;
    total_cost_usd: number;
    subscription_duration_days: number;
    average_daily_usage: {
      api_calls: number;
      ai_analyses: number;
    };
  };
  projections: {
    monthly_api_calls: number;
    monthly_ai_analyses: number;
    monthly_cost_usd: number;
    recommended_plan?: string;
  };
}

// Subscription Upgrade/Downgrade Types
export interface PlanChangeRequest {
  new_plan_id: number;
  change_type: 'upgrade' | 'downgrade';
  effective_date: 'immediate' | 'next_billing_cycle';
  proration_amount?: number;
  reason?: string;
}

export interface PlanChangePreview {
  current_plan: Plan;
  new_plan: Plan;
  change_type: 'upgrade' | 'downgrade';
  proration_amount: number;
  next_billing_amount: number;
  effective_date: Timestamp;
  feature_changes: {
    added: string[];
    removed: string[];
    modified: Array<{
      feature: string;
      old_value: any;
      new_value: any;
    }>;
  };
}

// Subscription Cancellation Types
export interface CancellationRequest {
  reason: string;
  feedback?: string;
  effective_date: 'immediate' | 'end_of_period';
  cancel_immediately: boolean;
}

export interface CancellationPreview {
  effective_date: Timestamp;
  refund_amount: number;
  access_until: Timestamp;
  data_retention_period: number; // days
  reactivation_deadline?: Timestamp;
}

// Subscription Renewal Types
export interface RenewalInfo {
  next_renewal_date: Timestamp;
  renewal_amount: number;
  auto_renew_enabled: boolean;
  payment_method_valid: boolean;
  renewal_warnings: string[];
}

// Subscription Discount Types
export interface Discount {
  id: string;
  code: string;
  type: 'percentage' | 'fixed_amount';
  value: number;
  description: string;
  valid_from: Timestamp;
  valid_until: Timestamp;
  max_uses?: number;
  current_uses: number;
  applicable_plans: number[];
  is_active: boolean;
}

export interface DiscountApplication {
  discount_id: string;
  subscription_id: UUID;
  applied_at: Timestamp;
  discount_amount: number;
  expires_at?: Timestamp;
}